core/src/main/java/com/sap/ai/sdk/core/AiCoreServiceWithDeployment.java

```diff
@@ -0,0 +1,179 @@
+package com.sap.ai.sdk.core;
+
+import com.sap.ai.sdk.core.client.DeploymentApi;
+import com.sap.ai.sdk.core.client.model.AiDeployment;
+import com.sap.cloud.sdk.cloudplatform.connectivity.DefaultHttpDestination;
+import com.sap.cloud.sdk.cloudplatform.connectivity.Destination;
+import com.sap.cloud.sdk.cloudplatform.connectivity.HttpDestination;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+
+/** Connectivity convenience methods for AI Core with deployment. */
+@RequiredArgsConstructor(access = AccessLevel.PUBLIC)
+public class AiCoreServiceWithDeployment implements AiCoreServiceStub {
+
+  // the deployment id to be used
+  @Nonnull private final Function<AiCoreServiceWithDeployment, String> deploymentId;
+
+  // the base destination to be used
+  @Nonnull private final Function<AiCoreServiceWithDeployment, Destination> destination;
+
+  // the resource group, "default" if null
+  @Nullable private final Function<AiCoreServiceWithDeployment, String> resourceGroup;
+
+  /**
+   * Create a new instance of the AI Core service with a specific deployment id and destination.
+   *
+   * @param deploymentId The deployment id handler.
+   * @param destination The destination handler.
+   */
+  public AiCoreServiceWithDeployment(
+      @Nonnull final Function<AiCoreServiceWithDeployment, String> deploymentId,
+      @Nonnull final Function<AiCoreServiceWithDeployment, Destination> destination) {
+    this(deploymentId, destination, null);
+  }
+
+  @Nonnull
+  @Override
+  public Destination destination() {
+    final var dest = destination.apply(this).asHttp();
+    final var builder = DefaultHttpDestination.fromDestination(dest);
+    updateDestination(builder, dest);
+    return builder.build();
+  }
+
+  /**
+   * Set the resource group.
+   *
+   * @param resourceGroup The resource group.
+   * @return A new instance of the AI Core service.
+   */
+  @Nonnull
+  public AiCoreServiceWithDeployment withResourceGroup(@Nonnull final String resourceGroup) {
+    return withResourceGroup((core) -> resourceGroup);
+  }
+
+  /**
+   * Set the resource group.
+   *
+   * @param resourceGroup The resource group handler.
+   * @return A new instance of the AI Core service.
+   */
+  @Nonnull
+  public AiCoreServiceWithDeployment withResourceGroup(
+      @Nonnull final Function<AiCoreServiceWithDeployment, String> resourceGroup) {
+    return new AiCoreServiceWithDeployment(deploymentId, destination, resourceGroup);
```

Comment:
This is really confusing!
Is this supposed to be public?
How even do you use it? Do you have a customer side example?

Reply:
* `AiCoreService` offers **lazy** evaluation of `destination()` and `client()`
  * Destinations are not resolved in advance. This avoid unnecessary compuation.
* `AiCoreServiceWithDeployment` is subclass of  `AiCoreService`
  * Due to lazy evaluation, the class stores **no literals** - only handlers.
  * The accessors and constructors allow for setting the handlers.
  * **Convenience methods** allow **for literals**, they are translated to handlers.
  * (Why `Function` instead of `Supplier`? Because some of those literals dynamically depend on destination and deployment.)

We definitely can make these methods `protected` to not interfere with auto-completion!
If users see necessity for workarounds they can still extend the class and access `protected` code.

Reply:
> How even do you use it? Do you have a customer side example?

Resolve a resource group id dynamically.

```diff
  private static final OrchestrationCompletionApi API =
      new OrchestrationCompletionApi(
          new AiCoreService()
              .withDeploymentByScenario("orchestration")
-             .withResourceGroup("default")
+             .withResourceGroup(OrchestrationController::getMyResourceGroupId)
              .client());

+ private static String getMyResourceGroupId(AiCoreServiceWithDeployment core) {
+   return new ResourceGroupApi(core.client())
+       .getAll().getResources().stream()
+           .filter(r -> "foobar-tenant".equalsIgnoreCase(r.getTenantId()))
+           .map(BckndResourceGroup::getResourceGroupId)
+           .findAny()
+           .orElseThrow();
+ }
```

Reply:
While thinking about it, I think you're right:
* Resource Group Id is not dependent on `deploymentId`.
* Also using the `destination` with "altered" path, any lookup will AI Core interaction will fail, e.g. `ResourceGroupApi`.
* If the user can neither leverage provided destination nor deployment, the handler itself becomes less convenient than just expecting a `String` literal.

Reply:
I defused the code.
```diff
  Function<AiCoreServiceWithDeployment, String> deploymentId

- Function<AiCoreServiceWithDeployment, Destination> destination
+ Supplier<Destination> destination

- Function<AiCoreServiceWithDeployment, String> resourceGroup
+ String resourceGroup
```